lib/kubernetes-deploy/render_task.rb

```diff
@@ -46,29 +40,38 @@ def run!(stream, only_filenames = [])
 
     private
 
-    def render_filenames(stream, filenames)
-      exceptions = []
-      @logger.phase_heading("Rendering template(s)")
-
-      filenames.each do |filename|
-        begin
-          render_filename(filename, stream)
-        rescue KubernetesDeploy::InvalidTemplateError => exception
-          exceptions << exception
-          log_invalid_template(exception)
+    def template_sets_paths(only_filenames)
+      if @template_paths.present?
+        # Validation will catch @template_paths & @template_dir being present
+        @template_paths
+      elsif only_filenames.blank?
+        [File.expand_path(@template_dir || '')]
+      else
+        absolute_template_dir = File.expand_path(@template_dir || '')
+        only_filenames.map do |name|
+          File.join(absolute_template_dir, name)
         end
       end
+    end
 
-      unless exceptions.empty?
-        raise exceptions[0]
+    def render_templates(stream, template_sets)
+      @logger.phase_heading("Rendering template(s)")
+      count = 0
+      template_sets.with_resource_definitions_and_filename(render_erb: true,
+          current_sha: @current_sha, bindings: @bindings, raw: true) do |rendered_content, filename|
+        render_filename(rendered_content, filename, stream)
+        count += 1
       end
+
+      count
+    rescue KubernetesDeploy::InvalidTemplateError => exception
+      log_invalid_template(exception)
+      raise
     end
 
-    def render_filename(filename, stream)
+    def render_filename(rendered_content, filename, stream)
```

Comment:
Minor point but this function is no longer doing what it says: the renderer has already run by the time we get here. Said that, we are _rendering_ to the stream, but that's muddying the waters IMO. Not sure what we should call it instead, though (`write_to_stream`?)